### PR TITLE
Add docker support for SkyPilot

### PR DESCRIPTION
**Pull request by cblmemo**
This PR enable user-provided docker image on VM.

TODO:
- Test SSHProxyCommand on real-world usecase

Tested (run the relevant ones):

- [x] Any manual or new tests for this PR (please specify below)
     - `pytest tests/test_smoke.py::test_job_queue_with_docker` 
     - `pytest tests/test_smoke.py::test_docker_preinstalled_package` 
- [x] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests:
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Review by Michaelvll — COMMENTED** [review]
Thanks a lot for adding the docker feature @cblmemo! I am still going through the PR, but would like to send out some of the comments first. The main concern I have here is that we are passing the `docker_user` all around. Is it possible that we just set the user in the docker image to be the same as the `ssh_user` to simplify the implementation?

File: sky/backends/backend_utils.py [hunk 1]
```diff
@@ -1106,9 +1124,18 @@ def wait_until_ray_cluster_ready(
...
     except RuntimeError as e:
         logger.error(e)
-        return False  # failed
+        return False, None  # failed
+
+    docker_user = None
+    if use_docker:
+        # pylint: disable=import-outside-toplevel
+        from sky.backends.docker_utils import docker_host_setup
+        docker_user = docker_host_setup(head_ip, cluster_config_file)
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/backends/backend_utils.py:1133]: Quick answer: this option requires the docker image containing a user named `my_username`, which is typically not the case for our default username `gcpuser` or `azureuser`. But this is the implementation previous to overriding Ray's `DockerCommandRunner`; since now we have the ability to modify the docker launch process, there may be some workarounds available and I'm trying it.

**Comment by cblmemo** [comment]
Some potential problems:
* Will docker reset after sky stop & sky start? If so, maybe mount some dir to host machine and tell user these dirs are persistent
* Will re-launch an up state docker cluster be influenced by port change to 10022?

File: sky/utils/command_runner.py [hunk 2]
```diff
@@ -138,14 +142,30 @@ def __init__(
...
-        self._ssh_proxy_command = ssh_proxy_command
+        if docker_user:
+            self.ip = 'localhost'
+            self.ssh_user = docker_user
+            if ssh_proxy_command is not None:
+                raise ValueError(
+                    'ssh_proxy_command is not supported if docker is used')
+            self._ssh_proxy_command = lambda ssh: ' '.join(
+                ssh + ssh_options_list(ssh_private_key, None
+                                      ) + ['-W', '%h:%p', f'{ssh_user}@{ip}'])
```
Comment (Michaelvll) [inline sky/utils/command_runner.py:161]: It seems possible to do multi-hop jump with `ssh`, with something like the following:
```
ssh -o ProxyCommand="ssh -W %h:%p user1@jump1" -o ProxyCommand="ssh -W %h:%p user2@jump2" user3@target
```

Is it possible for us to support it? We can also do it in another PR. : )

**Review by Michaelvll — COMMENTED** [review]
Thanks for adding this exciting feature @cblmemo! The code looks pretty good to me, and I am going to try it out soon. Left several comments for the code first.

File: sky/skylet/providers/azure/azure-config-template.json [hunk 3]
```diff
@@ -26,7 +26,7 @@
         "subnetName": "[concat('ray-', parameters('clusterId'), '-subnet')]"
     },
     "resources": [
-       {
+        {
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/skylet/providers/azure/azure-config-template.json:29]: No, it is probably the result of some auto formatter. Resolved now

File: sky/clouds/azure.py [hunk 4]
```diff
@@ -237,6 +239,18 @@ def make_deploy_resources_variables(
...
+        # This script will modify /etc/ssh/sshd_config and add a bash script
+        # into .bashrc. The bash script will restart sshd if it has not been
+        # restarted, identified by a file /tmp/__restarted is existing.
+        # pylint: disable=line-too-long
+        cloud_init_setup_commands = base64.b64encode(
+            textwrap.dedent("""\
+            #cloud-config
+                runcmd:
+                - sed -i 's/#Banner none/Banner none/' /etc/ssh/sshd_config
+                - echo '\\nif [ ! -f "/tmp/__restarted" ]; then\\n  sudo systemctl restart ssh\\n  sleep 2\\n  touch /tmp/__restarted\\nfi' >> /home/azureuser/.bashrc
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/clouds/azure.py:252]: IIRC it doesn't work if we restart the ssh service in the `runcmd` section here. And if we restart the service in .bashrc it works, so we need an indicator file to avoid restarting ssh service each time we run a command

[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/clouds/azure.py:252]: I'll do a test to double check this statement

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick update @cblmemo! I went through the code and it looks quite good to me. A problem I met is like the following:
```
sky launch -c test-docker test.yaml
```
```
resources:
  image_id: docker:continuumio/miniconda3

workdir: .

file_mounts:
  /target:
    source: gs://gcs-test-zhwu
    mode: MOUNT
  /examples: ./examples

setup: |
  echo hi
  ls /examples

run: |
  echo hi
  touch ~/user_file.txt
```

It seems the `sky launch` took up all the 8 CPU cores on my local machine, because it generates a significant amount of `/home/gcpuser/skypilot/sky/skylet/subprocess_daemon.py` process. Is it because of some unintended infinite loop or recursion?
```

File: sky/utils/command_runner.py [hunk 5]
```diff
@@ -87,14 +89,28 @@ def ssh_options_list(ssh_private_key: Optional[str],
         ssh_private_key,
     ] if ssh_private_key is not None else []
 
+    docker_proxy_command = [
+        '-o',
+        f'ProxyCommand={docker_ssh_proxy_command}',
+    ] if docker_ssh_proxy_command is not None else []
+
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/utils/command_runner.py:96]: The problem is we have 2 ProxyCommand shared same key (`ProxyCommand`) in the `arg_dict`, so the `docker_proxy_command` will directly cover what the original ProxyCommand is.

**Comment by cblmemo** [comment]
> Thanks for the quick update @cblmemo! I went through the code and it looks quite good to me. A problem I met is like the following:
> 
> ```
> sky launch -c test-docker test.yaml
> ```
> 
>  ```
> resources:
>   image_id: docker:continuumio/miniconda3
> 
> workdir: .
> 
> file_mounts:
>   /target:
>     source: gs://gcs-test-zhwu
>     mode: MOUNT
>   /examples: ./examples
> 
> setup: |
>   echo hi
>   ls /examples
> 
> run: |
>   echo hi
>   touch ~/user_file.txt
> ```
> 
> It seems the `sky launch` took up all the 8 CPU cores on my local machine, because it generates a significant amount of `/home/gcpuser/skypilot/sky/skylet/subprocess_daemon.py` process. Is it because of some unintended infinite loop or recursion?

This is fixed in [12b21d5](https://github.com/skypilot-org/skypilot/pull/1910/commits/12b21d56340698a507d5256bb45f4eeb5f83b3cc), where I forget to quote the docker proxy command before.

File: sky/clouds/azure.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/clouds/azure.py:252]: Yes, confirmed. Directly running `sudo systemctl restart ssh` doesn't work.

File: sky/skylet/log_lib.py [hunk 6]
```diff
@@ -260,13 +260,16 @@ def run_with_log(
...
+    # set -m to enable job control.
+    # Reference: https://stackoverflow.com/questions/11821378/what-does-bashno-job-control-in-this-shell-mean # pylint: disable=line-too-long
     # set -a is used for exporting all variables functions to the environment
     # so that bash `user_script` can access `conda activate`. Detail: #436.
     # Reference: https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html # pylint: disable=line-too-long
     # PYTHONUNBUFFERED is used to disable python output buffering.
     script = [
         textwrap.dedent(f"""\
             #!/bin/bash
+            set -m
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/skylet/log_lib.py]: I believe it is some intermediate stage of my development and it is not needed in the final version (I just done a test to confirm that). Removed now

File: sky/core.py [hunk 7]
```diff
@@ -499,17 +499,17 @@ def queue(cluster_name: str,
...
-    code = job_lib.JobLibCodeGen.get_job_queue(username, all_jobs)
-
     handle = backend_utils.check_cluster_available(
         cluster_name,
         operation='getting the job queue',
     )
+
     backend = backend_utils.get_backend_from_handle(handle)
+    all_jobs = not skip_finished
+    username: Optional[str] = handle.docker_user or getpass.getuser()
```
Comment (Michaelvll) [inline sky/core.py:509]: The username should be the local uesr who execute the job on the remote VM, instead of the remote user.

File: sky/utils/command_runner.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/utils/command_runner.py]: dict of list

File: sky/templates/azure-ray.yml.j2 [hunk 8]
```diff
@@ -111,25 +124,29 @@ setup_commands:
...
     sudo systemctl stop unattended-upgrades || true;
     sudo systemctl disable unattended-upgrades || true;
     sudo sed -i 's/Unattended-Upgrade "1"/Unattended-Upgrade "0"/g' /etc/apt/apt.conf.d/20auto-upgrades || true;
+    {%- endif %}
     p=$(mylsof "/var/lib/dpkg/lock-frontend"); echo "$p";
     sudo kill -9 `echo "$p" | tail -n 1` || true;
     sudo rm /var/lib/dpkg/lock-frontend;
     sudo pkill -9 dpkg;
     sudo pkill -9 apt-get;
     sudo dpkg --configure --force-overwrite -a;
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/templates/azure-ray.yml.j2:137]: maybe split to another PR

File: sky/clouds/azure.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/clouds/azure.py:252]: update: `systemctl restart ssh` also doesn't work

File: sky/templates/azure-ray.yml.j2 (hunk 8 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/templates/azure-ray.yml.j2]: Added now

**Review by Michaelvll — COMMENTED** [review]
This is fantastic @cblmemo! I am going through the code now.

Just to note down several optimizations we can do in the future (may be worth raising issues):
1. Use smaller AMIs/images for the cluster launched with docker image. As discovered by @suquark, our default cloud images are quite large and take a long time to provision. Since we are not using much of dependencies in the original image, we can use a smaller image to improve the provision speed.
2. Optimize the docker pull / or make the docker pull process visible in the progress bar. Currently, using a docker image can cause the `Launching` progress bar taking a long time, due to the `docker pull`. We can probably show something like `Luanching (pulling docker image)` to let the user now we are downloading the docker image.

File: sky/utils/log_utils.py [hunk 9]
```diff
@@ -92,6 +93,20 @@ def process_line(self, log_line):
...
+            self.state = self.ProvisionStatus.PULLING_DOCKER_IMAGES
+        if ('Status: Downloaded newer image' in log_line and
+                self.state == self.ProvisionStatus.PULLING_DOCKER_IMAGES):
+            self.status_display.stop()
+            logger.info(f'{colorama.Fore.GREEN}Docker image is downloaded.'
+                        f'{colorama.Style.RESET_ALL}')
+            self.status_display.start()
+            self.status_display.update(
+                '[bold cyan]Launching - Preparing SkyPilot runtime')
+            self.state = self.ProvisionStatus.RUNTIME_SETUP
```
Comment (Michaelvll) [inline sky/utils/log_utils.py:109]: Nice!

File: sky/skylet/providers/command_runner.py [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+        env_flags,
+        user_options_str,
+        '--net=host',
+        image,
+        'bash',
+    ]
+    return ' '.join(docker_run)
+
+
+class SkyDockerCommandRunner(DockerCommandRunner):
```
Comment (Michaelvll) [inline sky/skylet/providers/command_runner.py:64]: Can we annotate the palce where we update against original `DockerCommandRunner` with comments starting with `skypilot:`? Reason: in the future if we are going to update the command runner from the upstream, it will be easier to do that.

File: sky/utils/command_runner.py [hunk 11]
```diff
@@ -84,17 +90,32 @@ def ssh_options_list(ssh_private_key: Optional[str],
...
+        # If two proxy commands are specified, e.g.
+        # ssh -o ProxyCommand=cmd1 -o ProxyCommand=cmd2 ...
+        # then ssh will first ssh into the host using cmd2 (the later one) and
+        # then ssh from cmd2 host to the target host using cmd1. So, we put the
+        # docker proxy command at the beginning. This way, ssh will first
+        # establish a connection to the user-specified proxy host and then
+        # proceed to ssh from the proxy host to the docker host VM, and
+        # finally, ssh into the docker container.
+        arg_dict['ProxyCommand'].insert(0,
+                                        shlex.quote(docker_ssh_proxy_command))
```
Comment (Michaelvll) [inline sky/utils/command_runner.py:113]: nit: if we place this `if` block before the `if ssh_proxy_command is not None`, we can use `append` instead of `insert`?

File: sky/utils/log_utils.py [hunk 12]
```diff
@@ -76,14 +76,15 @@ class RayUpLineProcessor(LineProcessor):
...
+        PULLING_DOCKER_IMAGES = 2
 
     def __enter__(self):
         self.state = self.ProvisionStatus.LAUNCH
         self.status_display = safe_rich_status('[bold cyan]Launching')
         self.status_display.start()
 
     def process_line(self, log_line):
-        if ('Shared connection to' in log_line and
+        if ('Success.' in log_line and
```
Comment (Michaelvll) [inline sky/utils/log_utils.py:87]: Any reason we updated this?

File: sky/resources.py [hunk 13]
```diff
@@ -708,10 +709,29 @@ def _try_validate_local(self) -> None:
...
 
+        if self.extract_docker_image() is not None:
+            # TODO(tian): validate the docker image exists / of reasonable size
+            if self.accelerators is not None:
+                for acc in self.accelerators.keys():
+                    if acc.lower().startswith('tpu'):
+                        with ux_utils.print_exception_no_traceback():
+                            raise ValueError(
+                                'Docker image is not supported for TPU VM.')
+            return
```
Comment (Michaelvll) [inline sky/resources.py:733]: Should we check if the cloud specified supports the docker? For example, `sky launch --cloud lambda --image-id docker:xxx` will fail to recognize that the cloud does not support docker at the moment?

**Review by Michaelvll — COMMENTED** [review]
Thanks for the updates @cblmemo! This is awesome. Left some comments, mostly nits.

One question here is how we handle the failover after this PR is in? For example, `sky launch --image-id docker:xxx --gpus A100` will go to the lambda cloud first, should we add docker to the `Cloud.CloudImplementationFeatures` and disable those clouds that do not have the docker supported yet?

File: sky/utils/log_utils.py (hunk 12 above)
Comment (cblmemo) [reply sky/utils/log_utils.py:87]: Yes, the `Shared connection to` not always showed up in my test and it sometimes stuck at the `LAUNCH` state. There are two choices here: the `Success.` output by `updater.py`, or the `Got remote shell` output by `log_timer.py`.

```log
2023-07-12 16:32:20,102 SUCC updater.py:280 -- Success.
2023-07-12 16:32:20,102 INFO log_timer.py:30 -- NodeUpdater: ray-sky-e922-memory-head-76c3a3dc-compute: Got remote shell  [LogTimer=47191ms]
```

Comment (cblmemo) [reply sky/utils/log_utils.py:87]: Not sure if we need to create a new PR to do this

File: sky/clouds/azure.py [hunk 14]
```diff
@@ -237,6 +239,28 @@ def make_deploy_resources_variables(
...
+        # This script will modify /etc/ssh/sshd_config and add a bash script
+        # into .bashrc. The bash script will restart sshd if it has not been
+        # restarted, identified by a file /tmp/__restarted is existing.
+        # pylint: disable=line-too-long
+        cloud_init_setup_commands = base64.b64encode(
+            textwrap.dedent("""\
+            #cloud-config
+            runcmd:
+              - sed -i 's/#Banner none/Banner none/' /etc/ssh/sshd_config
+              - echo '\\nif [ ! -f "/tmp/__restarted" ]; then\\n  sudo systemctl restart ssh\\n  sleep 2\\n  touch /tmp/__restarted\\nfi' >> /home/azureuser/.bashrc
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/clouds/azure.py:252]: Yes, that doesn't work too

File: sky/backends/cloud_vm_ray_backend.py [hunk 15]
```diff
@@ -1811,6 +1812,7 @@ def need_ray_up(
             head_ip = None
             if len(ip_list) == 1:
                 head_ip = ip_list[0]
+            # If single node, docker_user is set up after handle is created.
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/backends/cloud_vm_ray_backend.py:1815]: deleted now

File: sky/backends/cloud_vm_ray_backend.py [hunk 16]
```diff
@@ -2221,6 +2225,26 @@ def get_hourly_price(self) -> float:
...
+            # Setup docker for all worker nodes. Head node setup is done
+            # in `backend_utils.wait_until_ray_cluster_ready`.
+            # Skip head ip here.
+            for worker_ip in ip_list[1:]:
+                docker_user = backend_utils.get_docker_user(
+                    worker_ip, cluster_config_file)
+                if self.docker_user is not None:
+                    assert docker_user == self.docker_user
+                else:
+                    self.docker_user = docker_user
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/backends/cloud_vm_ray_backend.py:2246]: Nice catch! 🙂 This appears to be a remnant of deprecated code. In the previous design, I need to run some setup commands in each worker and return the docker user in the same function, and now the setup command is no longer needed. Now we could just check the head IP.

File: sky/clouds/azure.py [hunk 17]
```diff
@@ -237,6 +239,28 @@ def make_deploy_resources_variables(
         gen_version = azure_catalog.get_gen_version_from_instance_type(
             r.instance_type)
         image_config = self._get_image_config(gen_version, r.instance_type)
+        # Setup commands to eliminate the banner and restart sshd.
+        # This script will modify /etc/ssh/sshd_config and add a bash script
+        # into .bashrc. The bash script will restart sshd if it has not been
+        # restarted, identified by a file /tmp/__restarted is existing.
+        # pylint: disable=line-too-long
+        cloud_init_setup_commands = base64.b64encode(
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/clouds/azure.py:247]: It seems Azure requires the custom data to be encoded. One exception I got:

```
Traceback (most recent call last):
  File "/tmp/skypilot_ray_up_mtv6p6sm.py", line 76, in <module>
    sdk.create_or_update_cluster('/home/memory/.sky/generated/sky-30dc-memory.yml', **{'no_restart': True})
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/sdk/sdk.py", line 48, in create_or_update_cluster
    use_login_shells=True,
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/_private/commands.py", line 289, in create_or_update_cluster
    no_monitor_on_head,
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/ray/autoscaler/_private/commands.py", line 726, in get_or_create_head_node
    provider.create_node(head_node_config, head_node_tags, 1)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/sky/skylet/providers/azure/node_provider.py", line 264, in create_node
    self._create_node(node_config, tags, count)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/sky/skylet/providers/azure/node_provider.py", line 314, in _create_node
    parameters=parameters,
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/azure/core/tracing/decorator.py", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/azure/core/polling/_poller.py", line 272, in wait
    raise self._exception  # type: ignore
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/azure/core/polling/_poller.py", line 189, in _start
    self._polling_method.run()
  File "/home/memory/install/miniconda3/envs/sky-dev/lib/python3.7/site-packages/azure/core/polling/base_polling.py", line 518, in run
    response=self._pipeline_response.http_response, error=err
azure.core.exceptions.HttpResponseError: (DeploymentFailed) At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.
Code: DeploymentFailed
Message: At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-deployment-operations for usage details.
Exception Details:      (BadRequest) {
          "error": {
            "code": "InvalidParameter",
            "message": "Custom data in OSProfile must be in Base64 encoding and with a maximum length of 87380 characters.",
            "target": "customData"
          }
        }
        Code: BadRequest
        Message: {
          "error": {
            "code": "InvalidParameter",
            "message": "Custom data in OSProfile must be in Base64 encoding and with a maximum length of 87380 characters.",
            "target": "customData"
          }
        }
```

File: sky/skylet/providers/command_runner.py (hunk 10 above)
Comment (cblmemo) [reply sky/skylet/providers/command_runner.py:64]: Added now

File: sky/resources.py (hunk 13 above)
Comment (cblmemo) [reply sky/resources.py:733]: fixed now

**Comment by cblmemo** [comment]
I've added the failover filter for lambda, scp, oci, and ibm. Now it could correctly chosen from three major clouds:

```
(sky-dev) ➜  skypilot git:(native-docker-support) ✗ sky launch --image-id docker:xxx --gpus A100
I 07-13 13:54:36 optimizer.py:636] == Optimizer ==
I 07-13 13:54:36 optimizer.py:648] Target: minimizing cost
I 07-13 13:54:36 optimizer.py:659] Estimated cost: $3.7 / hour
I 07-13 13:54:36 optimizer.py:659] 
I 07-13 13:54:36 optimizer.py:733] Considered resources (1 node):
I 07-13 13:54:36 optimizer.py:781] --------------------------------------------------------------------------------------------
I 07-13 13:54:36 optimizer.py:781]  CLOUD   INSTANCE        vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 07-13 13:54:36 optimizer.py:781] --------------------------------------------------------------------------------------------
I 07-13 13:54:36 optimizer.py:781]  GCP     a2-highgpu-1g   12      85        A100:1         us-central1   3.67          ✔     
I 07-13 13:54:36 optimizer.py:781] --------------------------------------------------------------------------------------------
I 07-13 13:54:36 optimizer.py:781] 
Launching a new cluster 'sky-98e0-memory'. Proceed? [Y/n]: n
Aborted!
```

```
(sky-dev) ➜  skypilot git:(native-docker-support) ✗ sky launch --image-id docker:xxx --gpus A100:8
I 07-13 13:54:44 optimizer.py:636] == Optimizer ==
I 07-13 13:54:44 optimizer.py:648] Target: minimizing cost
I 07-13 13:54:44 optimizer.py:659] Estimated cost: $27.2 / hour
I 07-13 13:54:44 optimizer.py:659] 
I 07-13 13:54:44 optimizer.py:733] Considered resources (1 node):
I 07-13 13:54:44 optimizer.py:781] --------------------------------------------------------------------------------------------------
I 07-13 13:54:44 optimizer.py:781]  CLOUD   INSTANCE              vCPUs   Mem(GB)   ACCELERATORS   REGION/ZONE   COST ($)   CHOSEN   
I 07-13 13:54:44 optimizer.py:781] --------------------------------------------------------------------------------------------------
I 07-13 13:54:44 optimizer.py:781]  Azure   Standard_ND96asr_v4   96      900       A100:8         eastus        27.20         ✔     
I 07-13 13:54:44 optimizer.py:781]  GCP     a2-highgpu-8g         96      680       A100:8         us-central1   29.39               
I 07-13 13:54:44 optimizer.py:781]  AWS     p4d.24xlarge          96      1152      A100:8         us-east-1     32.77               
I 07-13 13:54:44 optimizer.py:781] --------------------------------------------------------------------------------------------------
I 07-13 13:54:44 optimizer.py:781] 
Launching a new cluster 'sky-924a-memory'. Proceed? [Y/n]: n
Aborted!
```

File: sky/utils/command_runner.py [hunk 18]
```diff
@@ -74,7 +76,9 @@ def ssh_options_list(ssh_private_key: Optional[str],
         # Agent forwarding for git.
         'ForwardAgent': 'yes',
     }
-    if ssh_control_name is not None:
+    # SSH Control will have a severe delay when using docker_ssh_proxy_command.
+    # TODO(tian): Investigate why.
+    if ssh_control_name is not None and docker_ssh_proxy_command is None:
```
Comment (Michaelvll) [inline sky/utils/command_runner.py:81]: @landscapepainter

File: sky/utils/command_runner.py [hunk 19]
```diff
@@ -87,17 +92,31 @@ def ssh_options_list(ssh_private_key: Optional[str],
...
+        # must quote this value.
+        arg_dict['ProxyCommand'].append(shlex.quote(ssh_proxy_command))
+
     return ssh_key_option + [
         x for y in (['-o', f'{k}={v}']
-                    for k, v in arg_dict.items()
-                    if v is not None) for x in y
+                    for k, lv in arg_dict.items()
+                    for v in lv) for x in y
     ]
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/utils/command_runner.py]: This part of code is removed now

File: sky/clouds/ibm.py [hunk 20]
```diff
@@ -38,6 +38,10 @@ def _cloud_unsupported_features(
         return {
             clouds.CloudImplementationFeatures.CLONE_DISK_FROM_CLUSTER:
                 (f'Migrating disk is not supported in {cls._REPR}.'),
+            clouds.CloudImplementationFeatures.NATIVE_DOCKER_SUPPORT:
+                (f'Native docker is not supported in {cls._REPR}. '
+                 'You can try running docker command inside the '
+                 '`run` section in task.yaml.'),
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/clouds/ibm.py:44]: Because the `cloud-ray.yaml.j2` needs some tweaks to be properly executed in the docker

File: sky/resources.py [hunk 21]
```diff
@@ -730,6 +730,9 @@ def _try_validate_image_id(self) -> None:
                         with ux_utils.print_exception_no_traceback():
                             raise ValueError(
                                 'Docker image is not supported for TPU VM.')
+            if self.cloud is not None:
+                self.cloud.check_features_are_supported(
+                    {clouds.CloudImplementationFeatures.NATIVE_DOCKER_SUPPORT})
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/resources.py:697]: Maybe we could keep it since other features also failed on the resources is created 🤔 (such as ports)

File: sky/skylet/providers/command_runner.py [hunk 22]
```diff
@@ -172,10 +179,10 @@ def run_init(self, *, as_head: bool, file_mounts: Dict[str, str],
...
             docker_run_executed = True
 
-        # Setup Commands.
+        # SkyPilot: Setup Commands.
         # Most of docker images are using root as default user, so we set an
-        # alias for sudo to empty string, so any sudo in the following commands
-        # won't fail.
+        # alias for sudo to empty string, therefore any sudo in the following
+        # commands won't fail.
         # Disable apt-get from asking user input during installation.
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/skylet/providers/command_runner.py:186]: What did you refer to when saying 'assumptions'?

File: sky/resources.py [hunk 23]
```diff
@@ -670,10 +671,39 @@ def _try_validate_local(self) -> None:
...
+                for acc in self.accelerators.keys():
+                    if acc.lower().startswith('tpu'):
+                        with ux_utils.print_exception_no_traceback():
+                            raise ValueError(
+                                'Docker image is not supported for TPU VM.')
+            if self.cloud is not None:
+                self.cloud.check_features_are_supported(
+                    {clouds.CloudImplementationFeatures.NATIVE_DOCKER_SUPPORT})
+            if skypilot_config.get_nested(
+                (str(self.cloud).lower(), 'ssh_proxy_command'),
```
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/resources.py:699]: Humm, make sense... So we cannot determine whether the resources is valid or not until the cloud is chosen by the optimizer IIUC. Shall we just disable any cloud to have a proxy command when docker is used and no cloud is specified?

File: sky/skylet/providers/command_runner.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/providers/command_runner.py:186]: Assuming the `apt` is the package manager for the container?

File: sky/resources.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/resources.py:699]: We can probably skip the cloud in `Cloud.check_features_are_supported`, so we can skip the cloud with the `ssh_proxy_command` setup during the optimization.

**Review by Michaelvll — APPROVED** [review]
Thanks for the awesome feature @cblmemo! Just tested with 4 nodes with `launch/start/exec/stop/autostop/down`, all of them work like magic. Please feel free to merge this PR when the tests pass.

File: sky/skylet/providers/command_runner.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (cblmemo) [reply sky/skylet/providers/command_runner.py:186]: got it 🫡

**Comment by cblmemo** [comment]
All smoke tests passed with the latest commit. Merging to master now 🫡

**Comment by cblmemo** [comment]
Refactored the proxy command checking. @Michaelvll PTAL 🫡

**Review by Michaelvll — APPROVED** [review]
LGTM. Thanks for the quick update @cblmemo!
